src/utils/xhr-utils.js

```diff
@@ -0,0 +1,624 @@
+ * type being used in AMP runtime that needs serialization. The `Headers` type
+ * also needs serialization, but callers should not be passing `Headers`
+ * object in `init`, as that fails `fetchPolyfill` on browsers that don't
+ * support fetch. Some serialization-needing types for `init.body` such as
+ * `ArrayBuffer` and `Blob` are already supported by the structured clone
+ * algorithm. Other serialization-needing types such as `URLSearchParams`
+ * (which is not supported in IE and Safari) and `FederatedCredentials` are
+ * not used in AMP runtime.
+ *
+ * @param {string} input The URL of the XHR to convert to structured
+ *     cloneable.
+ * @param {!FetchInitDef} init The options of the XHR to convert to structured
+ *     cloneable.
+ * @return {{input: string, init: !FetchInitDef}} The serialized structurally-
+ *     cloneable request.
+ * @private
+ */
+export function toStructuredCloneable(input, init) {
+  const newInit = Object.assign({}, init);
+  if (isFormDataWrapper(init.body)) {
+    newInit.headers = newInit.headers || {};
+    newInit.headers['Content-Type'] = 'multipart/form-data;charset=utf-8';
+    newInit.body = fromIterator(init.body.entries());
+  }
+  return {input, init: newInit};
+}
+
+/**
+ * De-serializes a fetch response that was made possible to be passed to
+ * `postMessage()`, i.e., can be cloned using the
+ * [structured clone algorithm](http://mdn.io/Structured_clone_algorithm).
+ *
+ * The response is assumed to be serialized in the following way:
+ *
+ * 1. Transform the entries in the headers of the response into an
+ * `!Array<!Array<string>>` holding the list of header entries, where each
+ * element in the array is a header entry (key-value pair) represented as a
+ * 2-element array. The header key is case-insensitive.
+ *
+ * 2. Include the header entry list and `status` and `statusText` properties
+ * of the response in as `headers`, `status` and `statusText` properties of
+ * `init`.
+ *
+ * 3. Include the body of the response serialized as string in `body`.
+ *
+ * 4. Return a new object having properties `body` and `init`.
+ *
+ * The response is de-serialized in the following way:
+ *
+ * 1. If the `Response` type is supported and `responseType` is not
+ * document, pass `body` and `init` directly to the constructor of `Response`.
+ *
+ * 2. Otherwise, populate a fake XHR object to pass to `FetchResponse` as if
+ * the response is returned by the fetch polyfill.
+ *
+ * 3. If `responseType` is `document`, also parse the body and populate
+ * `responseXML` as a `Document` type.
+ *
+ * @param {JsonObject|string|undefined} response The structurally-cloneable
+ *     response to convert back to a regular Response.
+ * @param {string|undefined} responseType The original response type used to
+ *     initiate the XHR.
+ * @return {!FetchResponse|!Response} The deserialized regular response.
+ * @private
+ */
+export function fromStructuredCloneable(response, responseType) {
```

Comment:
@jridgewell this should be the exact copy as well, except a small change highlighted below

Reply:
not resolved.